File: dev/executor/readiness.ts
import { ProcessManager } from "./process-manager";
import { printViteReadyFallback } from "./vite-adapter";
import { runtimeStateFile } from "./paths";
import type { ManagedProcessState } from "./types";
import { fail, formatSince, printCommand, runText, sleepMs } from "./utils";
import {
	reloadExecutorService,
	verifyExecutorServiceActive,
	showExecutorServiceLogs,
	showInstanceLogs,
} from "./journal";

const pm = new ProcessManager(
	Object.fromEntries(
		Object.entries(process.env).map(([key, value]) => [key, value ?? ""]),
	),
);

function addressMatchesPort(address: string, port: string): boolean {
	return new RegExp(`:${port}$`).test(address);
}

const portConflictRe = /Port\s+\d+\s+is\s+already\s+in\s+use/i;

export interface RuntimeSnapshotView {
	state: ManagedProcessState;
	pid: number;
	startedAt: number | null;
	restartAttempts: number;
	nextRetryAt: number | null;
	lastExit: {
		exitCode: number | null;
		signalCode: string | null;
		at: number;
	} | null;
	lastError: string | null;
}

export interface RuntimeStateView {
	instances: Record<string, RuntimeSnapshotView>;
}

const managedProcessStates = new Set<ManagedProcessState>([
	"stopped",
	"starting",
	"running",
	"stopping",
	"backoff",
	"blocked",
]);

function isObject(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}

function parseRuntimeSnapshot(value: unknown): RuntimeSnapshotView | null {
	if (!isObject(value)) return null;
	if (
		typeof value.state !== "string" ||
		!managedProcessStates.has(value.state as ManagedProcessState) ||
		typeof value.pid !== "number" ||
		!Number.isInteger(value.pid) ||
		value.pid < 0
	) {
		return null;
	}

	const lastExit = isObject(value.lastExit)
		? {
				exitCode:
					typeof value.lastExit.exitCode === "number" ||
					value.lastExit.exitCode === null
						? value.lastExit.exitCode
						: null,
				signalCode:
					typeof value.lastExit.signalCode === "string" ||
					value.lastExit.signalCode === null
						? value.lastExit.signalCode
						: null,
				at:
					typeof value.lastExit.at === "number"
						? value.lastExit.at
						: 0,
			}
			: null;

	return {
		state: value.state as ManagedProcessState,
		pid: value.pid,
		startedAt:
			typeof value.startedAt === "number" || value.startedAt === null
				? value.startedAt
				: null,
		restartAttempts:
			typeof value.restartAttempts === "number" &&
			Number.isInteger(value.restartAttempts)
				? value.restartAttempts
				: 0,
		nextRetryAt:
			typeof value.nextRetryAt === "number" || value.nextRetryAt === null
				? value.nextRetryAt
				: null,
		lastExit,
		lastError: typeof value.lastError === "string" ? value.lastError : null,
	};
}

export async function readRuntimeState(): Promise<RuntimeStateView | null> {
	const file = Bun.file(runtimeStateFile);
	if (!(await file.exists())) return null;

	try {
		const raw = (await file.json()) as unknown;
		if (!isObject(raw) || !isObject(raw.instances)) return null;

		const instances: Record<string, RuntimeSnapshotView> = {};
		for (const [name, value] of Object.entries(raw.instances)) {
			const snapshot = parseRuntimeSnapshot(value);
			if (!snapshot) return null;
			instances[name] = snapshot;
		}
		return { instances };
	} catch {
		return null;
	}
}

export function runtimeStateFailure(
	name: string,
	label: string,
	snapshot: RuntimeSnapshotView,
): string | null {
	const detail = snapshot.lastError
		? `: ${snapshot.lastError}`
		: snapshot.lastExit
			? ` (exit ${snapshot.lastExit.exitCode ?? "unknown"}${
					snapshot.lastExit.signalCode
						? `, signal ${snapshot.lastExit.signalCode}`
						: ""
				})`
			: "";

	if (snapshot.state === "blocked") {
		return `[executor] ${label} blocked for ${name}${detail}`;
	}
	if (snapshot.state === "backoff") {
		return `[executor] ${label} failed for ${name}: instance is in backoff${detail}`;
	}
	if (snapshot.state === "stopped") {
		return `[executor] ${label} failed for ${name}: instance is stopped${detail}`;
	}
	return null;
}

function sameRuntimeSnapshot(
	left: RuntimeSnapshotView,
	right: RuntimeSnapshotView,
): boolean {
	return (
		left.state === right.state &&
		left.pid === right.pid &&
		left.startedAt === right.startedAt &&
		left.lastError === right.lastError &&
		left.lastExit?.at === right.lastExit?.at
	);
}

function runtimeStartObserved(
	snapshot: RuntimeSnapshotView,
	before: RuntimeSnapshotView | null,
	requireRestart: boolean,
): boolean {
	if (snapshot.state !== "running" || !processAlive(snapshot.pid)) return false;
	if (!requireRestart || !before || before.state !== "running") return true;
	return snapshot.pid !== before.pid || snapshot.startedAt !== before.startedAt;
}

function processAlive(pid: number): boolean {
	if (pid <= 0) return false;
	try {
		process.kill(pid, 0);
		return true;
	} catch (error) {
		return (error as NodeJS.ErrnoException).code === "EPERM";
	}
}

export function runtimeStartComplete(
	runtimeAvailable: boolean,
	snapshot: RuntimeSnapshotView | null,
	before: RuntimeSnapshotView | null,
	requireRestart: boolean,
): boolean {
	return (
		runtimeAvailable &&
		snapshot !== null &&
		runtimeStartObserved(snapshot, before, requireRestart)
	);
}

export function runtimeStopComplete(
	runtimeAvailable: boolean,
	snapshot: RuntimeSnapshotView | null,
	initialPid: number,
): boolean {
	return (
		runtimeAvailable &&
		(!snapshot || snapshot.state === "stopped") &&
		!processAlive(initialPid)
	);
}

function printReadinessResult(
	args: string[],
	result: ReturnType<typeof runText>,
): void {
	printCommand(args);
	const stdout = result.stdout.trimEnd();
	const stderr = result.stderr.trimEnd();
	if (stdout) console.log(stdout);
	if (stderr) console.error(stderr);

	const details: string[] = [];
	if (result.spawnError) details.push(String(result.spawnError));
	if (result.timedOut) details.push("timed out");
	if (result.outputOverflow) details.push("output limit exceeded");
	if (details.length > 0) {
		console.error(`[executor] readiness log query failed: ${details.join(", ")}`);
	}
}

async function stableRuntimeStart(
	name: string,
	label: string,
	before: RuntimeSnapshotView | null,
	requireRestart: boolean,
): Promise<boolean> {
	const runtime = await readRuntimeState();
	const snapshot = runtime?.instances[name];
	if (!snapshot) return false;

	const failure = runtimeStateFailure(name, label, snapshot);
	if (failure && (!before || !sameRuntimeSnapshot(snapshot, before))) {
		fail(failure);
	}
	if (
		!runtimeStartComplete(
			Boolean(runtime),
			snapshot,
			before,
			requireRestart,
		)
	) {
		return false;
	}

	await sleepMs(100);
	const confirmedRuntime = await readRuntimeState();
	const confirmed = confirmedRuntime?.instances[name];
	if (!confirmed) return false;
	const confirmedFailure = runtimeStateFailure(name, label, confirmed);
	if (
		confirmedFailure &&
		(!before || !sameRuntimeSnapshot(confirmed, before))
	) {
		fail(confirmedFailure);
	}
	return runtimeStartComplete(
		Boolean(confirmedRuntime),
		confirmed,
		before,
		requireRestart,
	);
}

async function waitForLogPattern(
	name: string,
	pattern: string,
	since: string,
	label: string,
	timeoutSeconds = 15,
	port?: string,
	before: RuntimeSnapshotView | null = null,
	requireRestart = false,
): Promise<void> {
	const args = [
		"/usr/bin/journalctl",
		"--user",
		"-t",
		`executor/${name}`,
		"--since",
		since,
		"--output=cat",
		"--no-pager",
	];
	const deadline = Date.now() + timeoutSeconds * 1_000;
	let conflictHandled = false;
	let logPatternSeen = false;

	while (Date.now() < deadline) {
		if (await stableRuntimeStart(name, label, before, requireRestart)) {
			return;
		}

		const result = runText(args);
		if (result.stdout.includes(pattern)) logPatternSeen = true;

		if (!conflictHandled && port && portConflictRe.test(result.stdout)) {
			conflictHandled = true;
			console.error(
				`[executor] port ${port} is occupied; refusing to kill an external process`,
			);
		}

		await sleepMs(1_000);
	}

	const result = runText(args);
	if (result.stdout.includes(pattern)) logPatternSeen = true;
	printReadinessResult(args, result);
	if (logPatternSeen) {
		console.error(
			`[executor] observed ${label} log output, but runtime state did not reach running`,
		);
	}
	const runtime = await readRuntimeState();
	const snapshot = runtime?.instances[name];
	const failure = snapshot && runtimeStateFailure(name, label, snapshot);
	if (failure) fail(failure);
	fail(`timed out waiting for ${label} on ${name}`);
}

async function waitForInstanceReady(
	name: string,
	pattern: string,
	since: string,
	label: string,
	cmd: string,
	dir: string,
	port: string,
	timeoutSeconds = 15,
): Promise<void> {
	const args = [
		"/usr/bin/journalctl",
		"--user",
		"-t",
		`executor/${name}`,
		"--since",
		since,
		"--output=cat",
		"--no-pager",
	];
	const startedAt = Date.now();
	const deadline = Date.now() + timeoutSeconds * 1_000;
	let portReadyAt = 0;
	let conflictHandled = false;

	while (Date.now() < deadline) {
		const runtime = await readRuntimeState();
		const snapshot = runtime?.instances[name];
		const failure = snapshot && runtimeStateFailure(name, label, snapshot);
		if (failure) fail(failure);
		const runtimeRunning =
			snapshot?.state === "running" && processAlive(snapshot.pid);

		const result = runText(args);
		if (runtimeRunning && result.stdout.includes(pattern)) {
			return;
		}

		if (!conflictHandled && port && portConflictRe.test(result.stdout)) {
			conflictHandled = true;
			console.error(
				`[executor] port ${port} is occupied; refusing to kill an external process`,
			);
		}

		if (runtimeRunning) {
			const addresses = await pm.listeningAddresses(name, cmd);
			if (addresses.some((address) => addressMatchesPort(address, port))) {
				if (!portReadyAt) {
					portReadyAt = Date.now();
				}

				if (Date.now() - portReadyAt >= 5_000) {
					await printViteReadyFallback(
						dir,
						name,
						port,
						Date.now() - startedAt,
					);
					return;
				}
			}
		}

		await sleepMs(portReadyAt ? 500 : 1_000);
	}

	const result = runText(args);
	printReadinessResult(args, result);
	const runtime = await readRuntimeState();
	const snapshot = runtime?.instances[name];
	const failure = snapshot && runtimeStateFailure(name, label, snapshot);
	if (failure) fail(failure);
	fail(`timed out waiting for ${label} on ${name}`);
}

export async function changeAndWait(
	name: string,
	cmd: string,
	dir: string,
	pattern: string,
	label: string,
	readyPattern?: string,
	readyPort?: string,
): Promise<void> {
	const since = formatSince();
	const before = (await readRuntimeState())?.instances[name] ?? null;
	reloadExecutorService();
	verifyExecutorServiceActive();
	await waitForLogPattern(
		name,
		pattern,
		since,
		label,
		15,
		readyPort,
		before,
		label === "reload",
	);
	if (readyPattern && readyPort) {
		await waitForInstanceReady(
			name,
			readyPattern,
			since,
			`${label} ready`,
			cmd,
			dir,
			readyPort,
		);
	} else if (readyPattern) {
		await waitForLogPattern(name, readyPattern, since, `${label} ready`);
	}
	showExecutorServiceLogs(since);
	showInstanceLogs(name, since);
}

export async function stopAndVerify(name: string): Promise<void> {
	const since = formatSince();
	const before = (await readRuntimeState())?.instances[name] ?? null;
	const initialPid = before?.pid ?? 0;

	reloadExecutorService();
	verifyExecutorServiceActive();

	showExecutorServiceLogs(since);

	const args = [
		"/usr/bin/journalctl",
		"--user",
		"-t",
		`executor/${name}`,
		"--since",
		since,
		"--output=cat",
		"--no-pager",
	];

	const deadline = Date.now() + 5_000;
	while (Date.now() < deadline) {
		const runtime = await readRuntimeState();
		const snapshot = runtime?.instances[name];

		if (snapshot?.state === "running") {
			const restarted =
				!before ||
				before.state !== "running" ||
				snapshot.pid !== before.pid ||
				snapshot.startedAt !== before.startedAt;
			if (restarted) {
				const result = runText(args);
				printReadinessResult(args, result);
				fail(`instance restarted after stop request: ${name}`);
			}
		}

		if (runtimeStopComplete(Boolean(runtime), snapshot ?? null, initialPid)) {
			showInstanceLogs(name, since);
			return;
		}

		const result = runText(args);
		const logs = result.stdout;
		if (logs.includes("[executor] starting ")) {
			printReadinessResult(args, result);
			fail(`instance restarted after stop request: ${name}`);
		}

		await sleepMs(1_000);
	}

	const result = runText(args);
	printReadinessResult(args, result);
	const runtime = await readRuntimeState();
	const snapshot = runtime?.instances[name];
	const detail = snapshot
		? ` (runtime state: ${snapshot.state}${
				snapshot.lastError ? `, ${snapshot.lastError}` : ""
			})`
		: "";
	showInstanceLogs(name, since);
	fail(`timed out waiting for instance to stop: ${name}${detail}`);
}

File: dev/executor/src/readiness.rs
use crate::cli::{fail, CliError};
use crate::journal::{
    reload_executor_service, show_executor_service_logs, show_instance_logs,
    verify_executor_service_active,
};
use crate::pm::{ManagedProcessSnapshot, ManagedProcessState, ProcessManager, ProcessManagerOptions};
use crate::runner::{format_since, print_command, run_text, sleep_ms, RunOptions, RunResult};
use crate::state::RuntimeStateFile;
use crate::vite::print_vite_ready_fallback;
use std::collections::BTreeMap;
use std::time::{Duration, Instant};

pub async fn read_runtime_state() -> Result<Option<RuntimeStateFile>, CliError> {
    crate::state::read_runtime_state().await
}

fn process_alive(pid: u32) -> bool {
    if pid == 0 {
        return false;
    }
    match nix::sys::signal::kill(nix::unistd::Pid::from_raw(pid as i32), None) {
        Ok(()) => true,
        Err(nix::errno::Errno::EPERM) => true,
        Err(_) => false,
    }
}

pub fn runtime_state_failure(
    name: &str,
    label: &str,
    snapshot: &ManagedProcessSnapshot,
) -> Option<String> {
    let detail = if let Some(error) = &snapshot.last_error {
        format!(": {error}")
    } else if let Some(exit) = &snapshot.last_exit {
        let mut text = format!(" (exit {}", exit.exit_code.map(|code| code.to_string()).unwrap_or_else(|| "unknown".into()));
        if let Some(signal) = &exit.signal_code {
            text.push_str(&format!(", signal {signal}"));
        }
        text.push(')');
        text
    } else {
        String::new()
    };
    match snapshot.state {
        ManagedProcessState::Blocked => Some(format!(
            "[executor] {label} blocked for {name}{detail}"
        )),
        ManagedProcessState::Backoff => Some(format!(
            "[executor] {label} failed for {name}: instance is in backoff{detail}"
        )),
        ManagedProcessState::Stopped => Some(format!(
            "[executor] {label} failed for {name}: instance is stopped{detail}"
        )),
        _ => None,
    }
}

fn same_runtime_snapshot(left: &ManagedProcessSnapshot, right: &ManagedProcessSnapshot) -> bool {
    left.state == right.state
        && left.pid == right.pid
        && left.started_at == right.started_at
        && left.last_error == right.last_error
        && left.last_exit.as_ref().map(|exit| exit.at) == right.last_exit.as_ref().map(|exit| exit.at)
}

fn runtime_start_observed(
    snapshot: &ManagedProcessSnapshot,
    before: Option<&ManagedProcessSnapshot>,
    require_restart: bool,
) -> bool {
    if snapshot.state != ManagedProcessState::Running || !process_alive(snapshot.pid) {
        return false;
    }
    if !require_restart || before.is_none_or(|before| before.state != ManagedProcessState::Running) {
        return true;
    }
    let before = before.unwrap();
    snapshot.pid != before.pid || snapshot.started_at != before.started_at
}

pub fn runtime_start_complete(
    runtime_available: bool,
    snapshot: Option<&ManagedProcessSnapshot>,
    before: Option<&ManagedProcessSnapshot>,
    require_restart: bool,
) -> bool {
    runtime_available
        && snapshot.is_some_and(|snapshot| runtime_start_observed(snapshot, before, require_restart))
}

pub fn runtime_stop_complete(
    runtime_available: bool,
    snapshot: Option<&ManagedProcessSnapshot>,
    initial_pid: u32,
) -> bool {
    runtime_available
        && snapshot.is_none_or(|snapshot| snapshot.state == ManagedProcessState::Stopped)
        && !process_alive(initial_pid)
}

fn print_readiness_result(args: &[String], result: &RunResult) {
    print_command(args);
    let stdout = result.stdout.trim_end();
    let stderr = result.stderr.trim_end();
    if !stdout.is_empty() {
        println!("{stdout}");
    }
    if !stderr.is_empty() {
        eprintln!("{stderr}");
    }
    let mut details = Vec::new();
    if result.spawn_error.is_some() {
        details.push(result.spawn_error.clone().unwrap());
    }
    if result.timed_out {
        details.push("timed out".into());
    }
    if result.output_overflow {
        details.push("output limit exceeded".into());
    }
    if !details.is_empty() {
        eprintln!(
            "[executor] readiness log query failed: {}",
            details.join(", ")
        );
    }
}

async fn stable_runtime_start(
    name: &str,
    label: &str,
    before: Option<&ManagedProcessSnapshot>,
    require_restart: bool,
) -> Result<bool, CliError> {
    let runtime = read_runtime_state().await?;
    let snapshot = runtime.as_ref().and_then(|runtime| runtime.instances.get(name));
    let Some(snapshot) = snapshot else {
        return Ok(false);
    };
    if let Some(failure) = runtime_state_failure(name, label, snapshot) {
        if before.is_none_or(|before| !same_runtime_snapshot(snapshot, before)) {
            return Err(fail(failure));
        }
    }
    if !runtime_start_complete(true, Some(snapshot), before, require_restart) {
        return Ok(false);
    }
    tokio::time::sleep(Duration::from_millis(100)).await;
    let confirmed_runtime = read_runtime_state().await?;
    let confirmed = confirmed_runtime
        .as_ref()
        .and_then(|runtime| runtime.instances.get(name));
    let Some(confirmed) = confirmed else {
        return Ok(false);
    };
    if let Some(failure) = runtime_state_failure(name, label, confirmed) {
        if before.is_none_or(|before| !same_runtime_snapshot(confirmed, before)) {
            return Err(fail(failure));
        }
    }
    Ok(runtime_start_complete(
        confirmed_runtime.is_some(),
        Some(confirmed),
        before,
        require_restart,
    ))
}

async fn wait_for_log_pattern(
    name: &str,
    pattern: &str,
    since: &str,
    label: &str,
    timeout_seconds: u64,
    port: Option<&str>,
    before: Option<&ManagedProcessSnapshot>,
    require_restart: bool,
) -> Result<(), CliError> {
    let args = vec![
        "/usr/bin/journalctl".into(),
        "--user".into(),
        "-t".into(),
        format!("executor/{name}"),
        "--since".into(),
        since.into(),
        "--output=cat".into(),
        "--no-pager".into(),
    ];
    let deadline = Instant::now() + Duration::from_secs(timeout_seconds);
    let mut conflict_handled = false;
    let mut log_pattern_seen = false;
    let port_conflict = regex_port_conflict();
    while Instant::now() < deadline {
        if stable_runtime_start(name, label, before, require_restart).await? {
            return Ok(());
        }
        let result = run_text(&args, RunOptions::default());
        if result.stdout.contains(pattern) {
            log_pattern_seen = true;
        }
        if !conflict_handled {
            if let Some(port) = port {
                if port_conflict(&result.stdout) {
                    conflict_handled = true;
                    eprintln!(
                        "[executor] port {port} is occupied; refusing to kill an external process"
                    );
                }
            }
        }
        sleep_ms(1_000);
    }
    let result = run_text(&args, RunOptions::default());
    if result.stdout.contains(pattern) {
        log_pattern_seen = true;
    }
    print_readiness_result(&args, &result);
    if log_pattern_seen {
        eprintln!(
            "[executor] observed {label} log output, but runtime state did not reach running"
        );
    }
    let runtime = read_runtime_state().await?;
    if let Some(snapshot) = runtime.as_ref().and_then(|runtime| runtime.instances.get(name)) {
        if let Some(failure) = runtime_state_failure(name, label, snapshot) {
            return Err(fail(failure));
        }
    }
    Err(fail(format!("timed out waiting for {label} on {name}")))
}

fn regex_port_conflict() -> impl Fn(&str) -> bool {
    |stdout: &str| {
        let lower = stdout.to_ascii_lowercase();
        lower.contains("port ") && lower.contains("is already in use")
    }
}

fn address_matches_port(address: &str, port: &str) -> bool {
    address.ends_with(&format!(":{port}"))
}

async fn wait_for_instance_ready(
    name: &str,
    pattern: &str,
    since: &str,
    label: &str,
    cmd: &str,
    dir: &str,
    port: &str,
    timeout_seconds: u64,
) -> Result<(), CliError> {
    let args = vec![
        "/usr/bin/journalctl".into(),
        "--user".into(),
        "-t".into(),
        format!("executor/{name}"),
        "--since".into(),
        since.into(),
        "--output=cat".into(),
        "--no-pager".into(),
    ];
    let started_at = Instant::now();
    let deadline = Instant::now() + Duration::from_secs(timeout_seconds);
    let mut port_ready_at: Option<Instant> = None;
    let mut conflict_handled = false;
    let pm = ProcessManager::new(BTreeMap::new(), ProcessManagerOptions::default());
    while Instant::now() < deadline {
        let runtime = read_runtime_state().await?;
        let snapshot = runtime.as_ref().and_then(|runtime| runtime.instances.get(name));
        if let Some(snapshot) = snapshot {
            if let Some(failure) = runtime_state_failure(name, label, snapshot) {
                return Err(fail(failure));
            }
        }
        let runtime_running = snapshot.is_some_and(|snapshot| {
            snapshot.state == ManagedProcessState::Running && process_alive(snapshot.pid)
        });
        let result = run_text(&args, RunOptions::default());
        if runtime_running && result.stdout.contains(pattern) {
            return Ok(());
        }
        if !conflict_handled && regex_port_conflict()(&result.stdout) {
            conflict_handled = true;
            eprintln!("[executor] port {port} is occupied; refusing to kill an external process");
        }
        if runtime_running {
            let addresses = pm.listening_addresses(name, cmd).await.unwrap_or_default();
            if addresses
                .iter()
                .any(|address| address_matches_port(address, port))
            {
                let ready_at = *port_ready_at.get_or_insert_with(Instant::now);
                if ready_at.elapsed() >= Duration::from_secs(5) {
                    print_vite_ready_fallback(dir, name, port, started_at.elapsed().as_millis() as u64);
                    return Ok(());
                }
            }
        }
        sleep_ms(if port_ready_at.is_some() { 500 } else { 1_000 });
    }
    let result = run_text(&args, RunOptions::default());
    print_readiness_result(&args, &result);
    let runtime = read_runtime_state().await?;
    if let Some(snapshot) = runtime.as_ref().and_then(|runtime| runtime.instances.get(name)) {
        if let Some(failure) = runtime_state_failure(name, label, snapshot) {
            return Err(fail(failure));
        }
    }
    Err(fail(format!("timed out waiting for {label} on {name}")))
}

pub async fn change_and_wait(
    name: &str,
    cmd: &str,
    dir: &str,
    pattern: &str,
    label: &str,
    ready_pattern: Option<&str>,
    ready_port: Option<&str>,
) -> Result<(), CliError> {
    let since = format_since();
    let before_state = read_runtime_state().await.ok().flatten();
    let before = before_state
        .as_ref()
        .and_then(|runtime| runtime.instances.get(name))
        .cloned();
    reload_executor_service()?;
    verify_executor_service_active()?;
    wait_for_log_pattern(
        name,
        pattern,
        &since,
        label,
        15,
        ready_port,
        before.as_ref(),
        label == "reload",
    )
    .await?;
    if let (Some(ready_pattern), Some(ready_port)) = (ready_pattern, ready_port) {
        wait_for_instance_ready(
            name,
            ready_pattern,
            &since,
            &format!("{label} ready"),
            cmd,
            dir,
            ready_port,
            15,
        )
        .await?;
    } else if let Some(ready_pattern) = ready_pattern {
        wait_for_log_pattern(
            name,
            ready_pattern,
            &since,
            &format!("{label} ready"),
            15,
            None,
            None,
            false,
        )
        .await?;
    }
    show_executor_service_logs(Some(&since));
    show_instance_logs(name, Some(&since));
    Ok(())
}

pub async fn stop_and_verify(name: &str) -> Result<(), CliError> {
    let since = format_since();
    let before_state = read_runtime_state().await.ok().flatten();
    let before = before_state
        .as_ref()
        .and_then(|runtime| runtime.instances.get(name))
        .cloned();
    let initial_pid = before.as_ref().map(|snapshot| snapshot.pid).unwrap_or(0);
    reload_executor_service()?;
    verify_executor_service_active()?;
    show_executor_service_logs(Some(&since));
    let args = vec![
        "/usr/bin/journalctl".into(),
        "--user".into(),
        "-t".into(),
        format!("executor/{name}"),
        "--since".into(),
        since.clone(),
        "--output=cat".into(),
        "--no-pager".into(),
    ];
    let deadline = Instant::now() + Duration::from_secs(5);
    while Instant::now() < deadline {
        let runtime = read_runtime_state().await?;
        let snapshot = runtime.as_ref().and_then(|runtime| runtime.instances.get(name));
        if let Some(snapshot) = snapshot {
            if snapshot.state == ManagedProcessState::Running {
                let restarted = before.as_ref().is_none_or(|before| {
                    before.state != ManagedProcessState::Running
                        || snapshot.pid != before.pid
                        || snapshot.started_at != before.started_at
                });
                if restarted {
                    let result = run_text(&args, RunOptions::default());
                    print_readiness_result(&args, &result);
                    return Err(fail(format!(
                        "instance restarted after stop request: {name}"
                    )));
                }
            }
        }
        if runtime_stop_complete(runtime.is_some(), snapshot, initial_pid) {
            show_instance_logs(name, Some(&since));
            return Ok(());
        }
        let result = run_text(&args, RunOptions::default());
        if result.stdout.contains("[executor] starting ") {
            print_readiness_result(&args, &result);
            return Err(fail(format!(
                "instance restarted after stop request: {name}"
            )));
        }
        sleep_ms(1_000);
    }
    let result = run_text(&args, RunOptions::default());
    print_readiness_result(&args, &result);
    let runtime = read_runtime_state().await?;
    let snapshot = runtime.as_ref().and_then(|runtime| runtime.instances.get(name));
    let detail = snapshot
        .map(|snapshot| {
            let error = snapshot
                .last_error
                .as_deref()
                .map(|error| format!(", {error}"))
                .unwrap_or_default();
            format!(" (runtime state: {:?}{error})", snapshot.state)
        })
        .unwrap_or_default();
    show_instance_logs(name, Some(&since));
    Err(fail(format!(
        "timed out waiting for instance to stop: {name}{detail}"
    )))
}
